simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs

```diff
@@ -1,275 +1,409 @@
+        /// <summary>
+        /// On add item event
+        /// </summary>
+        public event AddItemEventHandler AddItem;
+
+        /// <summary>
+        /// On remove item event
+        /// </summary>
+        public event RemoveItemEventHandler RemoveItem;
+
+        /// <summary>
+        /// On clear collection event
+        /// </summary>
+        public event ClearCollectionEventHandlder ClearCollection;
+
+        /// <summary>
+        /// On commit collection event
+        /// </summary>
+        public event CommitCollectionEventHandlder CommitCollection;
+
+        /// <summary>
+        /// On item added event
+        /// </summary>
+        public event ItemAddedEventHandler ItemAdded;
+
+        /// <summary>
+        /// On item removed event
+        /// </summary>
+        public event ItemRemovedEventHandler ItemRemoved;
+
+        /// <summary>
+        /// On item cleared event
+        /// </summary>
+        public event CollectionClearedEventHandlder CollectionCleared;
+
+        /// <summary>
+        /// On collection committed event
+        /// </summary>
+        public event CollectionCommittedEventHandlder CollectionCommitted;
+        #endregion
+
+        #region Fields
+        private IList<T> items;
+        private IList<T> newItems;
+        private IList<T> removedItems;
+        #endregion
+
+        #region Constructor
+        /// <summary>
+        /// Initialize new stateful iterable
+        /// </summary>
+        /// <param name="initialList">Initial list of items</param>
+        public StatefulCollection(IEnumerable<T> initialList)
+        {
+            items = new List<T>(initialList);
+            newItems = new List<T>();
+            removedItems = new List<T>();
+        }
+        #endregion
+
+        #region Public Methods
+
+        /// <summary>
+        /// Shows a preview of the commited collection data
+        /// </summary>
+        /// <returns>data after a possible commit</returns>
+        public ObservableCollection<T> GetAsObservableCollection()
+        {
+            ObservableCollection<T> preview = new ObservableCollection<T>(this.items);
+            foreach (var item in this.newItems)
+            {
+                preview.Add(item);
+            }
+            return preview;
+        }
+
+        /// <summary>
+        /// Add a new item.
+        /// If the item already exists - ignore it
+        /// If the item was removed, put it in the item list again
+        /// </summary>
+        /// <param name="item">Item to add</param>
+        public void Add(T item)
+        {
+            // If the item is not added to any list yet
+            if (removedItems.Contains(item))
+            {
+                var args = new AddItemEventArgs
+                {
+                    Item = item,
+                    ReaddFromRemovedItems = true
+                };
+                AddItem?.Invoke(this, args);
+
+                if (!args.Cancel)
```

Comment:
do you really need to check this ? you are creating the args yourself. you know which value the args has.

Reply:
What do you mean? I check whether the event is not null, Böttcher args...

Reply:
> What do you mean? I check whether the event is not null, **Böttcher** args...

iphone plays games again 😆 

i meant `if (!args.Cancel)`